Added functions for mean downsampling with and without raw sql queries

suryatho opened the pull request:
## Changes

Added capabilities for mean downsampling with raw sql query when getting data by run id

## Notes

Left code for tracing timing of get downsampled data call in data_service.rs

It seems like the data returned by the versions of the method that doesn't use a raw sql query and the one that does don't differ other than usually just one timestamp is different. The data returned by n-th point downsampling vs mean downsampling is different (oh really?). Looks like the mean downsampling does work from eye test, but I haven't written any tests for it. 

I am unsure of how to gather timing data because there is differences when I call the function the first time versus subsequent times, and even then it seems to fluctuate after the first call, so all the different methods (n-th vs. mean vs. mean with raw sql) didn't really show any differences to me. How should I actually test my changes?

## Test Cases

- Didn't add tests yet

## Screenshots

<img width="1756" height="504" alt="image" src="https://github.com/user-attachments/assets/d38d244c-ebc2-4702-b6ec-8aa10e8e421c" />

## To Do

_Any remaining things that need to get done_

- [x] Bench changes

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `package-lock.json` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #457

suryatho committed:
Added functions for mean downsampling with and without raw sql queries
Did not update test file
Added traces for time to perform downsampling queries
#457

suryatho commented:
Should the `get_data_by_run_id_with_auto_downsampling()` function also return the downsampling info in the [`data_controller`](https://github.com/Northeastern-Electric-Racing/Argos/blob/2aac071b6b3b1fc86cd8492504809d5e932280cc/scylla-server/src/controllers/data_controller.rs#L21-L67), or would that be weird ?

bracyw commented:
> Should the `get_data_by_run_id_with_auto_downsampling()` function also return the downsampling info in the [`data_controller`](https://github.com/Northeastern-Electric-Racing/Argos/blob/2aac071b6b3b1fc86cd8492504809d5e932280cc/scylla-server/src/controllers/data_controller.rs#L21-L67), or would that be weird ?

I think to answer both the question about timing and the question about whether or not to still use  `get_data_by_run_id_with_auto_downsampling()` in the data controller (if that's what your asking), I would say disable the info level print statements (do this by commenting out the tracing subscriber in main.rs), and then just print the time the query started and ended.

An easier probably better way would be just write two tests that loop over the service functions and collect the time taken from invocation of the function to return time, and take the average of like 1,000 calls or something. Database should have a lot of data, run the sim for like 15 minutes and then stop calypso and siren, before running your test. (If you want to you could also setup and impl testing data (like seed data for the integration_test.sh script)

That was a lot of yapping so just @ me in slack with any questions that come up from this.

suryatho committed:
Added file to do benchmarking on downsampling
#457

suryatho commented:
I just added a file that runs some benchmarks on the two different mean downsampling methods. Its a sad day but the output lowk makes sense. I also tried to analyze the sql query itself but I didn't really know what to do with the data from that.

<img width="878" height="252" alt="Screenshot 2025-11-03 at 12 06 05 AM" src="https://github.com/user-attachments/assets/ee511d04-b76d-4d15-a566-5c996159eaec" />

As you can see the normal implementation just using diesel and rust is significantly faster though we probably aren't seeing any of it on the front end. 
Just a note because I didn't include it in the out, there was only around 8k points for the segment temp data and 10k points for the accelerometer data (1 point per vs 3 points per) so I just fed in a higher sampling rate than it would normally do (it wouldn't even try downsampling in these cases). Maybe that affects the data enough to see in the benchmark but I don't really know if it would.

Oh yeah let me know what I should remove from this branch before it goes back into develop

bracyw commented:
> I just added a file that runs some benchmarks on the two different mean downsampling methods. Its a sad day but the output lowk makes sense. I also tried to analyze the sql query itself but I didn't really know what to do with the data from that.
> 
> <img alt="Screenshot 2025-11-03 at 12 06 05 AM" width="878" height="252" src="https://private-user-images.githubusercontent.com/63197592/508800373-ee511d04-b76d-4d15-a566-5c996159eaec.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI0NzQzNzQsIm5iZiI6MTc2MjQ3NDA3NCwicGF0aCI6Ii82MzE5NzU5Mi81MDg4MDAzNzMtZWU1MTFkMDQtYjc2ZC00ZDE1LWE1NjYtNWM5OTYxNTllYWVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTA3VDAwMDc1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg4NzA0ZjEwNWVlYzE2MTU0NmMyYjYyZmU4NjMyYjYyYjM0YzA1NWEwYmI2NjhkMGI0NGQ3OGExNzZiMzkzMWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Gd_098TUEDFJNI-B5dDqwoGrJWtvnGB337O9v1nQdME">
> As you can see the normal implementation just using diesel and rust is significantly faster though we probably aren't seeing any of it on the front end. Just a note because I didn't include it in the out, there was only around 8k points for the segment temp data and 10k points for the accelerometer data (1 point per vs 3 points per) so I just fed in a higher sampling rate than it would normally do (it wouldn't even try downsampling in these cases). Maybe that affects the data enough to see in the benchmark but I don't really know if it would.
> 
> Oh yeah let me know what I should remove from this branch before it goes back into develop

It seems like the first test shows a preference towards raw SQL and the second one is very strong on diesel. Was the mock data shutoff while you run, also I would recommend maybe running on a larger sample set the 10k. Just because the two tests are so different in timing. If you try a larger sample size and it's still very different, then we can just go with diesel. Just want to make sure everything is benchmarked evenly as we can before making a decision.

suryatho committed:
Bench data across all data types
#457

suryatho commented:
I added to the bench script to do a bench for each separate data type and write to a file so I can look at the data ([here](https://docs.google.com/spreadsheets/d/1VmgILM2M2Lfm6iOgVRqBt-gN0zmG10ATRtAGXCnzMUI/edit?usp=sharing)). I settled on 30 iterations for each data type with 5 warm-up runs (I don't know if that actually does anything), with 30 the whole benchmarking operation took 30 minutes and I didn't want to run it again. At the top of the spreadsheet I noted some important statistics:

<img width="345" height="248" alt="image" src="https://github.com/user-attachments/assets/5023e888-4645-49f4-a5c2-a9d5c300547a" />

It feels like it's contradicting itself at first, but what I think this means is that for more data points the raw SQL query handles it quicker, but for smaller ones, the diesel implementation works better. Also, based on the standard deviation data the SQL query seems to be more consistent but I also don't really know statistics and I don't know if how I am calculating standard deviation or interpreting it is even right.

Inside the code, there are currently two separate implementations of the `get_mean_downsampled_data()` in `data_service.rs`, so one should probably go. Let me know what else should be removed such as new dependencies.

bracyw commented on scylla-server/src/services/data_service.rs:
```diff
@@ -82,35 +85,136 @@ pub async fn add_many(
+        let mut sum_values: Vec<f32> = vec![0.0; min_values_len];
+        for d in chunk {
+            for (i, v) in d.values.iter().take(min_values_len).enumerate() {
+                if let Some(val) = v {
```
We should really never get a None version of the value here but if we do we should just print a warning message, or error message using the trace subscriber setup. (you can look at batching_loop in db_handler for an example)

bracyw commented on scylla-server/src/services/data_service.rs:
```diff
@@ -82,35 +85,136 @@ pub async fn add_many(
-    // Expand to other downsampling algorithms later
-    let downsampled: Vec<Data> = all_data
+#[derive(QueryableByName)]
+struct AggRow {
```
I know we talked about but you can delete the raw sql stuff

bracyw reviewed (changes requested):
I know we talked about on Thursday but thought I would just put changes needed to merge now.

suryatho committed:
Remove bench script and unused dependencies.
Clean code in `data_service.rs`
Add warn when reading none value from server while downsampling
#457
